src/artifact_providers/github.ts

```diff
@@ -49,48 +52,52 @@ export class GitHubArtifactProvider extends BaseArtifactProvider {
+   * Searched for the artifact with the given revision, paging
+   * through results if necessary.
    *
    * @param revision
-   * @param page
+   * @returns The artifact or null.
    */
-  protected async getRevisionArtifact(
-    revision: string,
-    revisionDate?: string,
-    page = 0,
-    tries = 0
-  ): Promise<ArtifactItem> {
+  protected async searchForRevisionArtifact(revision: string): Promise<ArtifactItem|null>  {
     const { repoName: repo, repoOwner: owner } = this.config;
     const per_page = 100;
 
-    // https://docs.github.com/en/free-pro-team@latest/rest/reference/actions#artifacts
-    const artifactResponse = (
-      await this.github.actions.listArtifactsForRepo({
+    this.logger.debug(
+      `Searching GitHub artifacts for ${owner}/${repo}, revision ${revision}`
+    );
+
+    let checkNextPage = true;
+    let revisionDate;
+    for (let page = 0; checkNextPage; page++) {
+      // https://docs.github.com/en/free-pro-team@latest/rest/reference/actions#artifacts
+      const artifactResponse = await this.github.actions.listArtifactsForRepo({
         owner: owner,
         repo: repo,
         per_page,
         page,
-      })
-    ).data;
-
-    const { artifacts } = artifactResponse;
-    this.logger.trace(`All available artifacts on page ${page}:`, artifacts);
-
-    // We need to find the most recent archive where name matches the revision.
-    // XXX(BYK): we assume the artifacts are listed in descending date order on
-    // this endpoint.
-    // There is no public documentation on this but the observed data and
-    // common-sense logic suggests that this is a reasonably safe assumption.
-    const foundArtifact = artifacts.find(
-      artifact => artifact.name === revision
-    );
-    if (foundArtifact) {
-      this.logger.trace(`Found artifact on page ${page}:`, foundArtifact);
-      return foundArtifact;
-    }
+      });
+
+      const { artifacts, total_count } = artifactResponse.data;
+      this.logger.trace(`All available artifacts on page ${page}:`, artifacts);
+
+      // We need to find the most recent archive where name matches the revision.
+      // XXX(BYK): we assume the artifacts are listed in descending date order on
+      // this endpoint.
+      // There is no public documentation on this but the observed data and
+      // common-sense logic suggests that this is a reasonably safe assumption.
+      const foundArtifact = artifacts.find(
+        artifact => artifact.name === revision
+      );
+      if (foundArtifact) {
+        this.logger.trace(`Found artifact on page ${page}:`, foundArtifact);
+        return foundArtifact;
+      }
+
+      if (total_count <= per_page * (page + 1)) {
+        this.logger.debug(`No more pages remaining`);
+        break;
+      }
 
-    let checkNextPage = false;
-    if (artifactResponse.total_count > per_page * (page + 1)) {
       if (revisionDate === undefined) {
         revisionDate = (
           await this.github.git.getCommit({
```

Comment:
**One subtle change with this refactor**

Previously the revision date would be passed to the recursive call to `getRevisionArtifact`. With this change, the `revisionDate` value is retrieved once per call to `searchForRevisionArtifact`.

IF we want to reduce the `getCommit` API calls we could do the following:

1. Getting the revision date once before calling `searchForRevisionArtifact` and pass it in as a value. The down side of this is I'd expect the majority of artifacts to be found in the first page of results.
1. Return the `revisionDate` from `searchForRevisionArtifact` and reuse it. The down side here is we have to handle multiple return values which just muddles the code a little.

Curious if anyone has strong opinions about this. I'm fine with how it is right now, it keeps the retry logic consistent, otherwise I'd probably opt for the first option, get the revision date and pass it in to each search call.

Reply:
another idea might be:

```js
class LazyDate {
    #date = null;
    constructor(github) {
        this.github = github
    }
    get date() {
        return this.#date || (this.#date = await this.github.whatever(...));
    }
}
```

(idk what I'm doing, I've never written a class in js)

Reply:
Very good point, so we could do:

```
export class GitHubArtifactProvider extends BaseArtifactProvider {
  public readonly github: Octokit;
  private revisionDates: Map<string, string>;

  public constructor(config: ArtifactProviderConfig) {
    super(config);
    this.github = getGitHubClient();
    this.revisionDates = new Map<string, string>();
  }
  
  private async getRevisionDate(revision: string): Promise<string> {
    const { repoName: repo, repoOwner: owner } = this.config;
    return this.revisionDates.get(revision) || (
      await this.github.git.getCommit({
        owner,
        repo,
        commit_sha: revision,
      })
    ).data.committer.date;
  }
```

Reply:
This was a great idea. 

I didn't go with a class because we should be able to rely on Javascript scopes for this. I used generics so can keep type-safety and it removes the need for `LazyDate`, `LazyCommit`, `Lazy...` if we ever want this for something else.